Set window icon on X11

kitt-cat opened the pull request:
Closes #30644

Many X11 environments expect a window icon to be supplied [as pixel data on a window property `_NET_WM_ICON`](https://specifications.freedesktop.org/wm-spec/1.4/ar01s05.html#id-1.6.13).

I confirmed this change fixes the icon in xfce4 for me, I think its likely it also fixes https://github.com/zed-industries/zed/issues/37961 but I haven't tested it.

## Questions
* [`image::RgbaImage` is exposed to the public API of gpui](https://github.com/zed-industries/zed/pull/40096/files#diff-318f166d72ad9476bd0a116446f5db3897fc1a4eb1d49aaf8105608bcf49ea53R1136). I would guess this is undesirable, but I wasn't sure of the best way to use gpui's native `Image` type..
* Currently [the icon is embedded into the binary](https://github.com/zed-industries/zed/pull/40096/files#diff-89af0b4072205c53b518aa977d6be48997e1a51fa4dbf06c7ddd1fec99fc510eR101). If this is undesirable, zed could alternatively implement [icon lookup](https://specifications.freedesktop.org/icon-theme-spec/latest/#icon_lookup) and try and find its icon from the system at runtime.

## Future work
* It might be nice to expose a `set_window_icon` method also (it could be used for example to show dirty state in the icon somehow), but I'm unfamiliar with what other platforms support and if this could be beyond X11 (there is a [wayland protocol](https://wayland.app/protocols/xdg-toplevel-icon-v1) though!).

Release Notes:

- Fixed missing window icon on X11

mgsloan commented:
Thanks for working on this!  It does seem good to set NET_WM_ICON.  I pushed a few misc fixes / improvements.

Is it necessary to resize the icons to 256x256?  I think I'd prefer to just leave them as 512x512 and leave it up to the DE to resize as needed.  Then the changes to `build.rs` could be skipped and this could be used:

```
            const BYTES: &[u8] = {
                let release_channel = option_env!("RELEASE_CHANNEL");
                match release_channel {
                    Some(release_channel) => match release_channel.as_bytes() {
                        b"stable" => include_bytes!("../resources/app-icon.png"),
                        b"preview" => include_bytes!("../resources/app-icon-preview.png"),
                        b"nightly" => include_bytes!("../resources/app-icon-nightly.png"),
                        _ => include_bytes!("../resources/app-icon-dev.png"),
                    },
                    _ => include_bytes!("../resources/app-icon-dev.png"),
                }
            };
```

Alternatively, if it is important to use that size, could add files like `app-icon@0.5x.png`.

> [image::RgbaImage is exposed to the public API of gpui](https://github.com/zed-industries/zed/pull/40096/files#diff-318f166d72ad9476bd0a116446f5db3897fc1a4eb1d49aaf8105608bcf49ea53R1136). I would guess this is undesirable, but I wasn't sure of the best way to use gpui's native Image type..

I think that's fine, `image::Frame` is already involved in gpui's public API.

> Zed could alternatively implement [icon lookup](https://specifications.freedesktop.org/icon-theme-spec/latest/#icon_lookup) and try and find its icon from the system at runtime.

Would this allow GPUI apps to always set their NET_WM_ICON to the same icon as their desktop file?  Would be quite cool if so - such a mechanism would need to do its work on the background thread.  However, this approach seems fine for now.

yara-blue reviewed (dismissed):
Please resolve the conflicts, extract a function in build.rs and then we can merge :)

if you want too feel free to include the change mentioned by @mizhka as well. But that can be a separate PR too.

kitt-cat committed:
Set window icon on X11

This only decodes the icon to RgbaImage once, instead of per window,
adds FreeBSD support for X11 window icon and moves icon processing to
separate function in build.rs

kitt-cat commented:
Hello again, happy to!

When I was resolving conflicts I noticed the `image` crate was made an optional dependency in 8c0b088366545e5050711066a4be666266f28110). This PR currently makes it a non-optional runtime dependency, but I think it could be made conditional on linux/freebsd only, if you think that would be worthwhile?

I'd be happy to squash the commits as well if that would be helpful.

yara-blue commented:
> I think it could be made conditional on linux/freebsd only,

I dont think its worth it for zed, given we use the image crate in a bunch of other places. Feel free to make it unconditional again!


> I'd be happy to squash the commits as well if that would be helpful.

no need, we always squash merge.

yara-blue committed:
fix build file on other platforms

yara-blue committed:
fix build file

yara-blue committed:
revert build file refactor for windows icon

cameron1024 reviewed (approved):
Thanks for the contribution 😁 

Looks good
